include/swift/Basic/Defer.h

```diff
@@ -30,8 +31,17 @@ namespace swift {
       Fn();
     }
   };
+
+  namespace detail {
+    struct DeferTask {};
+    template<typename F>
+    DoAtScopeExit<typename std::decay<F>::type> operator+(DeferTask, F&& fn) {
+      return DoAtScopeExit<typename std::decay<F>::type>(fn);
```

Comment:
`fn` has to be wrapped in `std::forward<F>(fn)` to take care of the forwarding reference.